src/jupyterlabcontroller/services/prepuller.py

```diff
@@ -0,0 +1,168 @@
+"""Prepull images to Kubernetes nodes."""
+
+import asyncio
+import os
+
+from aiojobs import Scheduler
+from kubernetes_asyncio.client import V1Container, V1PodSpec
+from structlog.stdlib import BoundLogger
+
+from ..models.domain.rspimage import RSPImage
+from ..models.v1.prepuller_config import PrepullerConfiguration
+from ..storage.k8s import K8sStorageClient
+from .image import ImageService
+
+
+class Prepuller:
+    """Prepull images to Kubernetes nodes.
+
+    There should be a singleton of this class in the lab controller process.
+    It runs as a background task, prepulling images to nodes based on the
+    information gathered by
+    `~jupyterlabcontroller.services.image.ImageService`.
+
+    Parameters
+    ----------
+    config
+        Configuration for the prepuller.
+    namespace
+        Namespace in which to put prepuller pods.
+    image_service
+        Service to query for image information. Currently, the background
+        refresh thread of the image service is also managed by this class.
+    k8s_client
+        Client for talking to Kubernetes.
+    logger
+        Logger for messages.
+    """
+
+    def __init__(
+        self,
+        *,
+        config: PrepullerConfiguration,
+        namespace: str,
+        image_service: ImageService,
+        k8s_client: K8sStorageClient,
+        logger: BoundLogger,
+    ) -> None:
+        self._image_service = image_service
+        self._k8s_client = k8s_client
+        self._logger = logger
+        self._namespace = namespace
+
+        self._scheduler = Scheduler()
+        self._running = False
+
+    async def start(self) -> None:
+        if self._running:
+            self._logger.info("Prepuller already running, cannot start again")
+            return
+        self._logger.info("Starting prepuller tasks")
+
+        # Wait for the image data to populate in the foreground. This ensures
+        # that we populate image data before FastAPI completes its startup
+        # event, and therefore before we start answering requests. That in
+        # turn means a more accurate health check, since until we have
+        # populated image data our API is fairly useless. (It also makes life
+        # easier for the test suite.)
+        await self._image_service.prepuller_wait()
+        await self._scheduler.spawn(self._prepull_loop())
+        self._running = True
+
+    async def stop(self) -> None:
+        if not self._running:
+            self._logger.info("Prepuller tasks were already stopped")
+        self._logger.info("Cancelling prepuller tasks")
+        await self._scheduler.close()
```

Comment:
Is closing a closed scheduler OK, or do you need to check the scheduler's status (or just return) if not self._running ?

Reply:
Good catch, it looks like you can't reuse a scheduler. Will fix that.